### DISCUSSION TITLE
Add a particle system (For road creation and active navigation)

### DISCUSSION THREAD

**Discussion by Abelkrijgtalles** [Abelkrijgtalles/mojang-maps#12] [Ideas]
With this system, you could navigate more easily. So instead of doing `/goto`, you would be doing `/startnavigation`. This could also be used to visualize roads, or even more stuff (I can't think of more, but maybe later I will)

**Comment by Abelkrijgtalles** [comment]
I've already done this a bit in the branch [dev/tryingoutparticles](https://github.com/Abelkrijgtalles/mojang-maps/tree/dev/tryingoutparticles). But I don't think I want to make a monorepo out of this

**Comment by samuelh2005** [comment]
# Suggestions for the particle system
## Better road creation

An easier way to create a road could be using a "wand" (like Worldedit) to select where the nodes should be. Left clicking a block may add it as a node and right clicking the block may remove it from the road (or do nothing if it is not a node).

Or new commands could be used to add and remove nodes to the road.

## What the node visualization could look like

The node blocks will be highlighted in particles and a particle line will be drawn between nodes.

An example of what this could look like:
![image](https://github.com/Abelkrijgtalles/mojang-maps/assets/41990982/b62ddeef-73b8-492d-a38f-5f644847c0eb)

This was done with the [CurveBuilding](https://github.com/kous500/CurveBuilding) plugin.

**Comment by Abelkrijgtalles** [comment]
Thank you! I will be testing this out right now

**Comment by Abelkrijgtalles** [comment]
Im developing this in [dev/betterroadcreation](https://github.com/Abelkrijgtalles/mojang-maps/tree/dev/betterroadcreation)

**Comment by Abelkrijgtalles** [comment]
Cool animation from YouTube tutorial (My own system was very buggy and just didn't work)

https://github.com/Abelkrijgtalles/mojang-maps/assets/56084910/85e24105-7513-409d-bc0a-ca63cd6ff050

**Reply by Abelkrijgtalles** [reply]
This is the implemented version

**Reply by samuelh2005** [reply]
Nice start.

**Reply by Abelkrijgtalles** [reply]
thx

**Comment by Abelkrijgtalles** [comment]
Searching for a clean name for the register road item. I wanted to do something with colors. Here is my first prototype:
![2024-02-02_23 33 00](https://github.com/Abelkrijgtalles/mojang-maps/assets/56084910/4c381415-94e3-4d4f-a7ec-9f5b041095c6)

**Reply by Abelkrijgtalles** [reply]
Epilepsy mode gone wrong:
![2024-02-02_23 41 45](https://github.com/Abelkrijgtalles/mojang-maps/assets/56084910/00a25eb6-5467-48af-8328-27409f78ae1d)

**Reply by Abelkrijgtalles** [reply]
Fixed it, but I thought it would look better in my head:
![image](https://github.com/Abelkrijgtalles/mojang-maps/assets/56084910/d1b85e3b-0924-4b69-8066-8ca41018915f)

**Reply by Abelkrijgtalles** [reply]
There we go, but I still think it isn't as good:
![image](https://github.com/Abelkrijgtalles/mojang-maps/assets/56084910/355a1844-4e79-4614-a418-33a6c722d652)

**Reply by Abelkrijgtalles** [reply]
Pretty good if I say so myself, but could be better:
![2024-02-02_23 47 01](https://github.com/Abelkrijgtalles/mojang-maps/assets/56084910/7378be01-f50d-41d9-b61f-0026d339ad7d)

**Reply by Abelkrijgtalles** [reply]
I think imma stick to that

**Reply by samuelh2005** [reply]
Colour look good, Maybe it could be called "Road creation tool"?

**Reply by Abelkrijgtalles** [reply]
yeah, that's actually a better name. I'm going to change it

**Reply by Abelkrijgtalles** [reply]
> Colour look good, Maybe it could be called "Road creation tool"?

Or maybe just "Road Creator"?

**Reply by Abelkrijgtalles** [reply]
I think that road creation tool is better

**Reply by Abelkrijgtalles** [reply]
screw road creator

**Reply by Abelkrijgtalles** [reply]
Here's is the updated name:
![image](https://github.com/Abelkrijgtalles/mojang-maps/assets/56084910/9982d1e4-0725-4efd-99e9-317e2285b57c)
And the message kinda looks cool:
![image](https://github.com/Abelkrijgtalles/mojang-maps/assets/56084910/e01377ac-683e-4274-ac5c-57baaeb3c596)

**Reply by Abelkrijgtalles** [reply]
I think this will be the final name, but it could always change

**Reply by Abelkrijgtalles** [reply]
I'm now gonna work on the system that saves the road from what you select

**Reply by samuelh2005** [reply]
Looks good so far, I should port this to Fabric.

**Reply by Abelkrijgtalles** [reply]
Or use something like architectury (idk if they support maven), so you also get forge and (maybe) also neoforge

**Reply by Abelkrijgtalles** [reply]
But I think that that has to be its own discussion.

**Comment by Abelkrijgtalles** [comment]
Here's pretty much the final version of the road saving:

https://github.com/Abelkrijgtalles/mojang-maps/assets/56084910/ad4e1e09-1136-4d03-b556-229e2c13c884

**Reply by samuelh2005** [reply]
Looks good.
It is a lot better than the current system, more user friendly.

**Reply by Abelkrijgtalles** [reply]
Thx

**Reply by Abelkrijgtalles** [reply]
I was thinking of just scrapping the old system

**Reply by samuelh2005** [reply]
That's probably a good idea now that we have the new one.

**Reply by Abelkrijgtalles** [reply]
I'm calling this a day and merging this to main as I think it's finished

**Reply by Abelkrijgtalles** [reply]
I'm now realizing how far I've come in 8 months. The first ever version was almost nothing, create roads (or even just locations) and travel to a specific road (or even older, see how you travel to each road)

**Reply by Abelkrijgtalles** [reply]
Back then I was just a 13-year-old kid with a funny not so realistic idea to create a whole navigation system using Dijkstra's algorithm. When I reach version 2.0, me from 8 months ago will be proud of me (my parents also, but they don't really understand the significance of this for me, so they don't like really care. I know they want to be, be it's just hard for them)

**Reply by Abelkrijgtalles** [reply]
But maybe with the more user-friendly side, they will understand the more significance of this. Idk.
